collect_app/src/test/java/org/odk/collect/android/widgets/BearingWidgetTest.java

```diff
@@ -43,4 +52,18 @@ public void setUp() throws Exception {
         super.setUp();
         barcodeData = RandomString.make();
     }
+
+    @Test
+    public void whenSensorUnavailable() {
+        BearingWidget bearingWidget = getWidget();
+        Button recordBearing = bearingWidget.getGetBearingButton();
+        when(bearingWidget.checkForRequiredSensors()).thenReturn(false);
```

Comment:
This doesn't do anything because it's too late! `checkForRequiredSensors()` is only called in the widget constructor. You could perhaps add a `protected` setter for `isSensorAvailable`.

The test works as expected when run on a device/emulator without the sensors.

Reply:
So, does it mean bearing widget doesn't need this test?

Reply:
No, it means that you have the right idea to want to make sure the widget acts as it does when no sensor is available but that the strategy doesn't work.

Reply:
Following on from what @lognaturel says, even if you commented out `when(bearingWidget.checkForRequiredSensors()).thenReturn(false);` the test will still pass.

Reply:
I've been having a think about how the issue could be solved and was exploring any clever ways to mock the method on the class before calling the constructor. But then it dawned on me that probably the easiest solution is to create a MockBearingWidget class inside BearingWidgetTest that simply overrides checkForRequiredSensors(). Then all that createWidget() widget has to do is return a new MockBearingWidget. The line in question can then be removed entirely.

example:

```
public class BearingWidgetTest extends BinaryWidgetTest<BearingWidget, StringData> {

    private class MockBearingWidget extends BearingWidget {

        public MockBearingWidget(Context context, FormEntryPrompt prompt) {
            super(context, prompt);
        }

        @Override
        public boolean checkForRequiredSensors() {
            return false;
        }
    }

...
...

    public BearingWidget createWidget() {
        return new MockBearingWidget(RuntimeEnvironment.application, formEntryPrompt);
    }

...

```

Reply:
I like that, @xsteelej! I had suggested adding a protected setter for `isSensorAvailable` but your approach is cleaner. 👍

Reply:
We could also move out `getGetBearingButton()` and `getEditTextAnswer()` out of BearingWidget and into MockBearingWidget, meaning that it BearingWidget is testable without any changes.